bin/BuildPackages.sh

```diff
@@ -52,86 +36,203 @@ build_packages() {
+if [[ -t 1 ]]
 then
+  COLOR=1
+else
+  COLOR=0
+fi
+
+# default LOGDIR
+LOGDIR=log
+
+# default LOGFILE
+LOGFILE=buildpackages
+
+# default FAILPKGFILE
+FAILPKGFILE=fail
+}
+
+# collect arguments
+collect_arguments(){
+while [[ "$#" -ge 1 ]]
+do
   case "$1" in
     --with-gaproot=*)
       GAPDIR=$(sed 's|--with-gaproot=||' <<< "$1")
       shift
     ;;
 
+    --with-gaproot)
+      shift
+      GAPDIR="$1"
+      shift
+    ;;
+
+    --with-logdir=*)
+      LOGDIR=$(sed 's|--with-logdir=||' <<< "$1")
+      shift
+    ;;
+
+    --with-logdir)
+      shift
+      LOGDIR="$1"
+      shift
+    ;;
+
+    --with-logfile=*)
+      LOGFILE=$(sed 's|--with-logfile=||' <<< "$1")
+      shift
+    ;;
+
+    --with-logfile)
+      shift
+      LOGFILE="$1"
+      shift
+    ;;
+
+    --with-failpkgfile=*)
+      FAILPKGFILE=$(sed 's|--with-failpkgfile=||' <<< "$1")
+      shift
+    ;;
+
+    --with-failpkgfile)
+      shift
+      FAILPKGFILE="$1"
+      shift
+    ;;
+
+    --no-color|--no-colors|--no-colour|--no-colours|\
+    --nocolor|--nocolors|--nocolour|--nocolours)
+      COLOR=0
+      shift
+    ;;
+
+    -*)
+      # Colors may not be defined at this point,
+      # then the default is to use coloring.
+      warning "Discarding unrecognized argument $1"
+      shift
+    ;;
+
     *)
-      cd ..
-      GAPDIR="$(pwd)"
-      cd "$CURDIR"
+      # Any other argument is considered to be a package directory to build.
+      PACKAGES+=("$1")
+      shift
     ;;
   esac
-fi
-if [[ -z "$GAPDIR" ]]
-then
-  pushd ..
-  GAPDIR="$(pwd)"
-  popd
-fi
-notice "Using GAP location: $GAPDIR"
+done
+}
 
 # We need to test if $GAPDIR is right
+test_GAPDIR(){
 if ! [[ -f "$GAPDIR/sysinfo.gap" ]]
 then
   error "$GAPDIR is not the root of a gap installation (no sysinfo.gap)" \
         "Please provide the absolute path of your GAP root directory as" \
         "first argument with '--with-gaproot=' to this script."
 fi
+}
 
+# set build flags for 32 bit
+set_build_flags_for_32_bit(){
 if [[ "$(grep -c 'ABI_CFLAGS=-m32' $GAPDIR/Makefile)" -ge 1 ]]
 then
   notice "Building with 32-bit ABI"
   ABI32=YES
   CONFIGFLAGS="CFLAGS=-m32 LDFLAGS=-m32 LOPTS=-m32 CXXFLAGS=-m32"
 fi
+}
 
 # packages to build
-if [[ "$#" -ge 1 ]]
+set_packages(){
+if [[ "${PACKAGES[0]}" = "" ]]
 then
-  # last arguments are packages to build
-  PACKAGES=("$@")
-else
-  # if there are no arguments, then build all packages
+  # If there were no extra arguments, therefore PACKAGES[0] is not defined,
+  # then build all packages.
   shopt -s nullglob
   PACKAGES=(*)
   shopt -u nullglob
 fi
+}
 
 # Many package require GNU make. So use gmake if available,
 # for improved compatibility with *BSD systems where "make"
 # is BSD make, not GNU make.
+set_make(){
 if ! [[ "x$(which gmake)" = "x" ]]
 then
   MAKE=gmake
 else
   MAKE=make
 fi
+}
+
+
+# print notices in green
+notice() {
+  if [[ "$COLOR" = "1" ]]
+  then
+    printf "\033[32m%s\033[0m\n" "$@"
+  else
+    printf "%s\n" "$@"
+  fi
```

Comment:
Why exactly  do you test the COLOR flag insider of the functions, instead of defining the functions differently in each case?